src/orion/core/io/database/ephemeraldb.py

```diff
@@ -0,0 +1,440 @@
+
+        """
+        found_documents = []
+        for document in self._documents:
+            if document.match(query):
+                found_documents.append(document.select(selection))
+
+        return found_documents
+
+    def _validate_index(self, document):
+        """Validate index values of a document
+
+        :raises: :exc:`DuplicateKeyError`: if the document contains unique indexes which are already
+        present in the database.
+        """
+        for index, values in self._indexes.items():
+            document_values = document.select({key: 1 for key in index})
+            if document_values in values:
+                raise DuplicateKeyError(
+                    "Duplicate key error: index={} value={}".format(index, document_values))
+
+    def _register_keys(self, document):
+        """Register index values of a new document"""
+        for index, values in self._indexes.items():
+            values.append(document.select({key: 1 for key in index}))
+
+    def _get_new_id(self):
+        """Return max id + 1"""
+        if self._documents:
+            return max(d['_id'] for d in self._documents) + 1
+
+        return 1
+
+    def insert_many(self, documents):
+        """Add new documents in the collection.
+
+        If the documents do not have a keys `_id`, they are assigned by default
+        the max id + 1.
+
+        :raises: :exc:`DuplicateKeyError`: if the document contains unique indexes which are
+            already present in the database.
+        """
+        for document in documents:
+            if '_id' not in document:
+                document['_id'] = self._get_new_id()
+            ephemeral_document = EphemeralDocument(document)
+            self._validate_index(ephemeral_document)
+            self._documents.append(ephemeral_document)
+            self._register_keys(ephemeral_document)
+
+        return True
+
+    def update_many(self, query, update):
+        """Update documents or upsert if not found.
+
+        If the document is not found, a new document which is the merge of query and update will be
+        inserted in the database.
+
+        :raises: :exc:`DuplicateKeyError`: if the update creates a duplication of unique indexes in
+            the database.
+        """
+        updates = 0
+        for document in self._documents:
+            if document.match(query):
+                document.update(update)
+                updates += 1
+
+        if not updates:
+            self._upsert(query, update)
+
+        return True
+
+    def _upsert(self, query, update):
+        """Insert the document when query was not found.
+
+        If update contains `$set`, then the new document is the combination of query and
+        update['$set'], otherwise the new document is `update`.
+        """
+        if "$set" in update:
+            new_document = copy.deepcopy(update["$set"])
+            new_document.update(query)
```

Comment:
This is what happens inside `mongodb`, so I mimicked the behavior since this is what we currently have with `MongoDB.write()`.

Ref: https://docs.mongodb.com/manual/reference/method/db.collection.update/#upsert-behavior